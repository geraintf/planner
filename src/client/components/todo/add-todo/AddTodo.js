import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Input, Button } from 'reactstrap';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faPlus from '@fortawesome/fontawesome-free-solid/faPlus';
import { encodeString } from 'src/utils/string-utils';

import './AddTodo.scss';

export default class AddTodo extends Component {
  static propTypes = {
    addTodo: PropTypes.func.isRequired
  };

  state = {
    value: ''
  };

  handleAdd = (event) => {
    const { value } = this.state;

    if (value !== '') {
      if (
        (event.type === 'keypress' && event.key === 'Enter')
        || event.type === 'click'
      ) {

        this.props.addTodo(encodeString(value));
        this.setState({
          value: ''
        });
      }
    }
  };

  handleOnChange = (event) => this.setState({ value: event.target.value });

  render() {
    return (
      <div className="add-todo">
        <Input
          placeholder="What do you need to get done?"
          onKeyPress={this.handleAdd}
          onChange={this.handleOnChange}
          value={this.state.value}
        />
        <Button
          color="secondary"
          onClick={this.handleAdd}
        >
          <FontAwesomeIcon icon={faPlus} />
        </Button>
      </div>
    )
  }
}

