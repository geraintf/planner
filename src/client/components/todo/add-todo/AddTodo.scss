
.add-todo {
  position: relative;
  opacity: 0.5;
  transition: opacity 200ms ease-in;

  &:hover {
    transition: opacity 100ms ease-in;
    opacity: 1;
  }

  button {
    border-radius: 0;
    position: absolute;
    top: 0;
    right: 0;

    &:focus {
      box-shadow: none;
      opacity: 1;
    }
  }

  .form-control {
    border-radius: 0;

    &:focus {
      border: 1px solid #ced4da;
      box-shadow: none;
    }
  }
}