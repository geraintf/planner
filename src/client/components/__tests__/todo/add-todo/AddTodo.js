import React from 'react';
import { shallow } from 'enzyme';
import { Input, Button } from 'reactstrap';

import AddTodo from '../../../todo/add-todo/AddTodo';

describe('AddTodo', () => {
  let output;
  let props;
  let addTodo;

  beforeEach(() => {
    addTodo = jest.fn();

    props = {
      addTodo
    };

    output = shallow(<AddTodo {...props} />);
  });

  it('should render correctly', () => {
    expect(output).toMatchSnapshot();
  });

  describe('given input is empty', () => {
    it('should not call addTodo', () => {
      output.setState({ value: '' });

      const input = output.find(Input);

      input.simulate('keypress', { type: 'keypress', key: 'x' });

      expect(addTodo).not.toHaveBeenCalled();
    });
  });

  describe('given the input is not empty', () => {
    describe('and a key is pressed', () => {
      describe('and the key is ENTER', () => {
        it('should call addTodo', () => {
          output.setState({ value: 'hey' });

          const input = output.find(Input);

          input.simulate('keypress', { type: 'keypress', key: 'Enter' });

          expect(addTodo).toHaveBeenCalledWith('hey');
          expect(output.state('value')).toEqual('');
        });
      });
    });

    describe('and the value in the input changes', () => {
      it('should update the state and the value of the input', () => {
        output.setState({ value: 'hey' });

        const input = output.find(Input);

        input.simulate('change', { target: { value: 'hey!' } });

        expect(output.state('value')).toEqual('hey!');
      });
    });

    describe('and the button is clicked', () => {
      it('should call addTodo', () => {
        output.setState({ value: 'hey' });

        const button = output.find(Button);

        button.simulate('click', { type: 'click' });

        expect(addTodo).toHaveBeenCalledWith('hey');
        expect(output.state('value')).toEqual('');
      });
    });
  });
});
