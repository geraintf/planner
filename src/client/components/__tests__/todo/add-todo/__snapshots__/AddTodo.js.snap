// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo should render correctly 1`] = `
<div
  className="add-todo"
>
  <Input
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="What do you need to get done?"
    type="text"
    value=""
  />
  <Button
    color="secondary"
    onClick={[Function]}
    tag="button"
  >
    <FontAwesomeIcon$1
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f067",
            "M448 294.2v-76.4c0-13.3-10.7-24-24-24H286.2V56c0-13.3-10.7-24-24-24h-76.4c-13.3 0-24 10.7-24 24v137.8H24c-13.3 0-24 10.7-24 24v76.4c0 13.3 10.7 24 24 24h137.8V456c0 13.3 10.7 24 24 24h76.4c13.3 0 24-10.7 24-24V318.2H424c13.3 0 24-10.7 24-24z",
          ],
          "iconName": "plus",
          "prefix": "fas",
        }
      }
      listItem={false}
      mask={null}
      name=""
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      transform={null}
    />
  </Button>
</div>
`;
